### PR TITLE
feat: add support for enterprise plans

### PR DESCRIPTION
**Pull request by fontanierh**

### CODE REVIEW

**Comment by fontanierh** [comment]
Also, when on the enterprise plan, the "subscription" page doesn't offer the usual "manage your plan" stuff.

**Review by spolu — COMMENTED** [review]
As per IRL my inclination would be:
- Have a new model for SubscriptionInvitation with a secret and a status
- No diff on workspace status, no diff on lib/auth.ts (modulo fixing the Promise<Promise<>>)
- Generate a link in Poke that can be shared over email
- Display same UX in plan page but based on SubscriptionInvitation instead of pendingPlan

Sorry for the big ask :grimacing:

File: front/lib/auth.ts [hunk 1]
```diff
@@ -152,11 +162,12 @@ export class Authenticator {
...
 
-    if (!user || !user.isDustSuperUser) {
-      return new Authenticator(workspace, user, "none", plan);
-    }
-
-    return new Authenticator(workspace, user, "admin", plan);
+    return new Authenticator({
+      workspace,
+      user,
+      role: !user?.isDustSuperUser ? "none" : "admin",
```
Comment (spolu) [inline front/lib/auth.ts:168]: nit `role: user?.isDustSuperUser ? "admin" : "none"`

File: front/lib/plans/stripe.ts [hunk 2]
```diff
@@ -30,13 +30,13 @@ async function getPriceId(productId: string): Promise<string | null> {
  * Once the users has completed the checkout, we will receive an event on our Stripe webhook
  */
 export const createCheckoutSession = async ({
-  owner,
+  workspaceSId,
```
Comment (spolu) [inline front/lib/plans/stripe.ts:33]: We generally call this `workspaceId` in the rest of the code

Comment (spolu) [reply front/lib/plans/stripe.ts:33]: But that being said the call site has access to the `owner: WorkspaceType` why not pass it, this cleaner IMHO?

File: front/pages/poke/[wId]/index.tsx [hunk 3]
```diff
@@ -128,11 +137,16 @@ export const getServerSideProps: GetServerSideProps<{
     slackbotEnabled = botEnabledRes.value.botEnabled;
   }
 
+  const planInvitation = await getPlanInvitation({
+    workspaceSId: workspace.sId,
```
Comment (spolu) [inline front/pages/poke/[wId]/index.tsx:141]: Can we use owner here as well?

File: front/pages/upgrade-enterprise/[secret].tsx [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+  });
+
+  if (!workspace) {
+    return {
+      notFound: true,
+    };
+  }
+
+  const checkoutUrl = await internalUpgradeWorkspace({
+    workspaceSId: workspace.sId,
```
Comment (spolu) [inline front/pages/upgrade-enterprise/[secret].tsx:42]: Can we use owner: WorkspaceType here as well?

**Review by spolu — COMMENTED** [review]
Looks great

as per IRL remove `internalUpgradeWorkspace` and accept to run `uipgradeWorkspac`  without user and role none.

Maybe rename it to something more explicit since what it does is not upgrade but generate a checkout link

**Comment by spolu** [comment]
Will PTAL post these changes :+1:

**Review by spolu — DISMISSED** [review]
LGTM :+1: Thanks for bearing with me on this one :pray:

**Comment by PopDaph** [comment]
Sorry for the rebase 🙈

File: front/lib/plans/stripe.ts [hunk 5]
```diff
@@ -30,18 +32,23 @@ async function getPriceId(productId: string): Promise<string | null> {
...
   stripeCustomerId,
 }: {
-  owner: WorkspaceType;
+  auth: Authenticator;
   planCode: string;
   productId: string;
   billingType: PaidBillingType;
   stripeCustomerId: string | null;
 }): Promise<string | null> => {
+  const workspace = auth.workspace();
```
[the remark this answers is not in the document]
Comment (fontanierh) [reply front/lib/plans/stripe.ts:47]: Oups yes, will rename 👍

**Review by spolu — DISMISSED** [review]
:100:

**Review by PopDaph — APPROVED** [review]
LGTM! Thanks 🤩 🔥

**Pull request merged**